[Nullability Annotations to Java Classes] Replace `findViewById` with `ViewBinding` for Notifications (`relatively safe`)

ParaskP7 opened the pull request:
Parent #18906

This PR's main focus is to replace `findViewById ` with `ViewBinding` for the [NotificationsDetailActivity.java](https://github.com/wordpress-mobile/WordPress-Android/blob/5b0d3f39c8cfe9c08b4e1547e46bc947f9be6b8f/WordPress/src/main/java/org/wordpress/android/ui/notifications/NotificationsDetailActivity.java#L88) class.

Additionally, this PR is also dealing with adding any missing nullability annotations to this activity class, focusing mainly on its fields, and, any effected neighbour classes. But first, all existing warnings on this class are being resolved/suppressed so as to make it easier to deal with any missing nullability checks (with help from the IDE), especially when a `@Nullable` annotation is added.

FYI: This change is `relatively safe`, meaning that although there are compile-time changes associated with this change, and it needs testing, the changes included in this PR are very targeted and specific to one screen, and one screen only. Thus, if this screen is tested appropriately, it should be safe to merge this to `trunk`.

-----

@mkevins let me know how I did on this PR, that is, in terms of using a nullable `binding` field and the additional warning and nullability related changes I made right after, just to make this class null proof. If you also feel that this is a good way to proceed with such PRs, I'll then use this pattern across the board and follow a similar approach on every such PR.

-----

`findViewById ` -> `ViewBinding`

1. [Replace findviewbyid with viewbinding for notifications detail](https://github.com/wordpress-mobile/WordPress-Android/commit/3a376ee66bb9038f687bd124b16fde6606d16b66)

Nullability Annotation List:

1. [Add missing nullability annotations to sdk override methods](https://github.com/wordpress-mobile/WordPress-Android/commit/720bc07c414fafd9d6236672ef9ce1c18b3324c4)
2. [Replace org jetbrains with androidx non null annotations](https://github.com/wordpress-mobile/WordPress-Android/commit/36308b1ddf80bf3de7580d8020e8ee04a985f83d)
3. [Add missing nullability annotations to app override methods](https://github.com/wordpress-mobile/WordPress-Android/commit/935b7e515c1a7619cd930d313434ee3e5bbf1400)
4. [Add missing nullability annotations to note id field](https://github.com/wordpress-mobile/WordPress-Android/commit/a94295900e131c71f69b121c9beca8516d7f5f77)
5. [Add missing nullability annotations to context parameters](https://github.com/wordpress-mobile/WordPress-Android/commit/fd5b3b1860b778194d2b6785ada6c7d07ffb8ce0)
6. [Add missing nullability annotations to change listener field](https://github.com/wordpress-mobile/WordPress-Android/commit/2da778e6612d7355e15b2cbf7f41c7510f4a8b60)
7. [Add missing nullability annotations to adapter field](https://github.com/wordpress-mobile/WordPress-Android/commit/8ccc1ab9a47a597ee75d5aa3a187b016563a0303)

Warnings Resolution List:

1. [Resolve unnecessary return warning on notifications detail](https://github.com/wordpress-mobile/WordPress-Android/commit/108accbae620a9402354c50ba924d47a349ee788)

Warnings Suppression List:

1. [Suppress unchecked warning on notifications detail ](https://github.com/wordpress-mobile/WordPress-Android/commit/91babcb07d6adf914a44f8434ce36ab965d9224c)
2. [Suppress deprecation warnings on notifications detail](https://github.com/wordpress-mobile/WordPress-Android/commit/f59a4b9491653763c7a595e47cfa1d1608d20b39)

-----

## To test:

1. Quickly smoke test any notification related functionality on both, the WordPress and Jetpack apps, and see if everything is working as expected.
8. In addition to the above smoke test, you can expand the below and follow the inner and more explicit test steps within:

<details>
    <summary>Notifications Details Screen [NotificationsDetailActivity.java]</summary>

ℹ️ This test applies to the Jetpack app.

- Go to the `Notifications` tab and tap on any notification.
- Verify that the `Notification Details` screen is shown and functioning as expected.

</details>

-----

## Regression Notes

1. Potential unintended areas of impact

    - N/A

3. What I did to test those areas of impact (or what existing automated tests I relied on)

    - See `To test` section above.

9. What automated tests I added (or what prevented me from doing so)

    - N/A

-----

## PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## UI Changes testing checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

ParaskP7 committed:
Refactor: Replace findviewbyid with viewbinding for notifications detail

ParaskP7 committed:
Analysis: Add missing nullability annotations to sdk override methods

ParaskP7 committed:
Analysis: Replace org jetbrains with androidx non null annotations

ParaskP7 committed:
Analysis: Add missing nullability annotations to app override methods

This change was applied to the 'onModerateCommentForNote(...)' method on
the 'OnNoteCommentActionListener' interface within the 'CommentActions'
class, and then got propagated to every class that is either
implementing this interface (ie. 'NotificationsDetailActivity') or
utilizing this interface (ie. 'CommentDetailFragment').

ParaskP7 committed:
Analysis: Resolve unnecessary return warning on notifications detail

Warning Message: "'return' is unnecessary as the last statement in a
'void' method"

ParaskP7 committed:
Analysis: Suppress unchecked warning on notifications detail

Warning Message: "Unchecked cast: 'java.lang.Object' to
'java.util.ArrayList<org.wordpress.android.models.Note>'"

ParaskP7 committed:
Analysis: Suppress deprecation warnings on notifications detail

Warning Messages:
- "'org.wordpress.android.ui.comments.CommentActions'
is deprecated"
- "'org.wordpress.android.ui.comments.CommentDetailFragment'
is deprecated"
- "'androidx.fragment.app.FragmentStatePagerAdapter'
is deprecated"

ParaskP7 committed:
Analysis: Add missing nullability annotations to note id field

This change was applied to the 'mNoteId' field on
'NotificationsDetailActivity' but was also propagated to the
'NotificationsUpdateServiceStarter.startService(...)' and
'NotificationsTable.getNoteById(...)' static methods.

ParaskP7 committed:
Analysis: Add missing nullability annotations to context parameters

ParaskP7 committed:
Analysis: Add missing nullability annotations to change listener field

ParaskP7 committed:
Analysis: Add missing nullability annotations to adapter field

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-Android into refactor/replace-findviewbyid-with-viewbinding-for-notifications

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/notifications/NotificationsDetailActivity.java

mkevins commented on WordPress/src/main/java/org/wordpress/android/ui/notifications/NotificationsDetailActivity.java:
```diff
@@ -97,22 +94,22 @@ public class NotificationsDetailActivity extends LocaleAwareActivity implements
+    @Nullable private ViewPager.OnPageChangeListener mOnPageChangeListener;
+    @Nullable private NotificationDetailFragmentAdapter mAdapter;
+
+    @Nullable private NotificationsDetailActivityBinding mBinding = null;
```
I'm not suggesting a change here, but just commenting to say that this feels like a good candidate for conversion to Kotlin (though out of scope here 😅). It's unfortunate that we don't have some equivalent of `lateinit` that we could use to avoid all the null guards 🤷‍♂️ . Save that for another day. 🙂

ParaskP7 replied:
Yea, I know @mkevins , I had the same feeling while doing this update, but then again I am not best friends with `lateinit` as people have long been starting to avoid using it whenever possible. Then again, on this specific `binding` case, and as per Google's [guidance on activities](https://developer.android.com/topic/libraries/view-binding#activities), using `lateinit` is safe.

FYI: The same does not apply to Google's [guidance on fragments](https://developer.android.com/topic/libraries/view-binding#fragments), since a nullable `binding` is being used there instead. Thus, using a nullable binding for Java activities, until at least we convert that to Kotlin, still sits okay with me, at least for now... 🤷 

PS: About converting to Kotlin, I wouldn't go that far, not only because it might be considered out of scope, but also, because although it might seems easy, a class with almost 700 lines of code deserves a Kotlinization PR of itself.

mkevins replied:
We can do our _own_ thing that is even _better_ than the guidance code. For example, I would replace their usage of the not-null assertion operator (`!!`) with a `checkNotNull`:

```kotlin
private var _binding: ResultProfileBinding? = null
// This property is only valid between onCreateView and
// onDestroyView.
private val binding get() = _binding!!
```

becomes:

```kotlin
private var _binding: ResultProfileBinding? = null
private val binding get() = checkNotNull(_binding) {
  "This property is only valid between onCreateView and onDestroyView."
}
```
 
And this would result in a more appropriate exception (`IllegalStateException`) and trace. We can do the same in Java, of course, but in Kotlin, this is idiomatic and elegant, whereas, in Java, it requires more effort (both to read and write such code).

ParaskP7 replied:
👋 @mkevins !

> We can do our own thing that is even better than the guidance code. For example, I would replace their usage of the not-null assertion operator (!!) with a checkNotNull:

We actually already decided to go another route, to only use the nullable `binding` and not go the `private val binding get() = _binding!!` road.

For more usages of this pattern, search for: `private var binding\: .*Binding\? \= null`

I think that so far this works well and as excepted, without us having much problem with using the `?` operator when needed to access that nullable `binding`, that is, since we try to use extension functions as much as possible in the meanwhile. For example, see `PreviewImageFragment.kt` and the fact that `binding?` is only used during `onDestroyView() { ... }`.

Having said that, comparing `private val binding get() = _binding!!` and the `private val binding get() = checkNotNull(_binding) { ... }`, I indeed like your approach better. And we can definitely revisit if we want to use a nullable `binding`, or a non null going forward, that is, if we find enough value in doing so.

Still, it seems to me all that is (kind of) out of scope for this Java nullability project we are currently undergoing and might only come handy when we start with the Kotlinization project, if and when, right? 🤔

> And this would result in a more appropriate exception (IllegalStateException) and trace. We can do the same in Java, of course, but in Kotlin, this is idiomatic and elegant, whereas, in Java, it requires more effort (both to read and write such code).

True! 👍

mkevins commented on WordPress/src/main/java/org/wordpress/android/ui/notifications/NotificationsDetailActivity.java:
```diff
@@ -581,31 +589,37 @@ public void onEventMainThread(NotificationEvents.NotificationsRefreshError error
             mNoteList = (ArrayList<Note>) notes.clone();
         }
 
+        @NonNull
```
I understand that the super method `getItem` is annotated as `@NonNull`, but the return value here is nullable, since `getDetailFragmentForNote` is nullable. I'm not sure why the compiler doesn't complain, but if the `Note` passed in as an argument is `null`, this will return `null`. Maybe we can assert that the note is not null? I.e. something like:

```diff
diff --git a/WordPress/src/main/java/org/wordpress/android/ui/notifications/NotificationsDetailActivity.java b/WordPress/src/main/java/org/wordpress/android/ui/notifications/NotificationsDetailActivity.java
index 1fcb956270..6bf0414f9e 100644
--- a/WordPress/src/main/java/org/wordpress/android/ui/notifications/NotificationsDetailActivity.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/notifications/NotificationsDetailActivity.java
@@ -68,6 +68,7 @@ import org.wordpress.android.widgets.WPViewPagerTransformer;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Objects;
 
 import javax.inject.Inject;
 
@@ -622,7 +623,8 @@ public class NotificationsDetailActivity extends LocaleAwareActivity implements
         @NonNull
         @Override
         public Fragment getItem(int position) {
-            return getDetailFragmentForNote(mNoteList.get(position));
+            @NonNull Note note = Objects.requireNonNull(mNoteList.get(position), "Note cannot be null");
+            return getDetailFragmentForNote(note);
         }
 
         @Override
```

Wdyt?

ParaskP7 replied:
> I understand that the super method getItem is annotated as @NonNull, but the return value here is nullable, since getDetailFragmentForNote is nullable.

Hmmm, interesting find @mkevins ! 💯 

> I'm not sure why the compiler doesn't complain, but if the Note passed in as an argument is null, this will return null.

The compiler doesn't complain because it doesn't know whether `getDetailFragmentForNote(...)` is `@Nullable` or `@NonNull`. So, if you annotate that method as `@Nullable` it will not start to complain. AFAIA, this is expected. 👍 

> Maybe we can assert that the note is not null? I.e. something like:

Thank you for suggesting this change! 🥇 

Thinking about it, maybe it is better to declare that this method accept a `@NonNull Note note` and returns a `@NonNull Fragment` instead, then delete the whole `if (note == null) { return null; }` "unnecessary" guarding code. The result will be the same, doing that or having `requireNonNull(...)` we will get a NPE either way, but this other way it makes it explicit that we don't want `note` to be nullable code-wise, instead of having a comment dictating it, which we know, devs might miss. Wdyt? 🤔

ParaskP7 replied:
Btw, I want to note here as an FYI, that this work here is related to adding missing nullability annotations to view related `fields` first (via the `findViewById` to `ViewBinding` conversion), and then maybe any other left-over fields missing such nullability annotation.

As such, the scope of this PR wasn't to deal with adding missing nullability annotations on method parameters or return values. Thus, you didn't see me deal with this issue you reported in this comment. Also, AFAIA, we might have more such issues.

Going forward with these `findViewById` to `ViewBinding` conversion PRs, I would like to continue that path, only focusing on the `fields` side of the problem and ignore adding any missing nullability annotations on the `method` side. I want to do that so as to not explode the scope of a PR and achieve merging it to `trunk` faster.

Do you agree with this approach @mkevins ? 🤔

mkevins replied:
> maybe it is better to declare that this method accept a @NonNull Note note and returns a @NonNull Fragment instead, then delete the whole if (note == null) { return null; } "unnecessary" guarding code.

Since this method is private, and it's only usage is here, I think this is a good approach. 👍 

>  instead of having a comment dictating it, which we know, devs might miss. Wdyt?

I'm not sure what you refer to by "comment", but I assume you mean the _message_ as the second argument to `Objects.requireNonNull`. The main advantage is that this helps pinpoint the NPE, and the reason. I'm not concerned about developers "missing" it, since it's intended for better stack traces (hard to miss there 😄). Even with the above suggested changes, we still need to assert that `mNoteList.get(position)` is `@NonNull`, so I think having the message is useful there. This, imo, is the _whole point_ of fail-fast. We want to fail as early (and as clearly) as possible when we are breaking contract.

ParaskP7 replied:
👋 @mkevins !

> Since this method is private, and it's only usage is here, I think this is a good approach. 👍

👍 

> I'm not sure what you refer to by "comment", but I assume you mean the message as the second argument to Objects.requireNonNull. The main advantage is that this helps pinpoint the NPE, and the reason. I'm not concerned about developers "missing" it, since it's intended for better stack traces (hard to miss there 😄).

Ah, for some reason I read the above as:

```java
@NonNull Note note = Objects.requireNonNull(mNoteList.get(position)); // Note cannot be null
```

🤦 😊 🤦 

> Even with the above suggested changes, we still need to assert that mNoteList.get(position) is @NonNull, so I think having the message is useful there. This, imo, is the whole point of fail-fast. We want to fail as early (and as clearly) as possible when we are breaking contract.

Oh yes, you are right about that and I was thinking about it too, was trying to figure out if `mNoteList.get(position)` will be null or not, that is, given the fact (or assumption) that a `Note` added to that list should never be null. But, I wanted to defer that investigation to another time.

Thus, I would still avoid adding the extra `Objects.requireNonNull(mNoteList.get(position), "Note cannot be null")` check there, because that would mean that this is now becoming the rule here, and we should strive doing that on every such instance of `Fragment getItem(int position) { ... }` in our project. But, I am not convinced yet that it is actually needed, I need to investigate if there are cases where `getItem(...)` will be called on the adapter but the list to get that item from might be empty, or else not having that item on that exact position, or else if that item can be null (so far my assumption is no to all of those questions). Does that make sense? 🤔

Also, this is another reason why I didn't want to go that deep in trying to solve every nullable problem while dealing with this `findViewById` to `ViewBinding` replacement, and only wanting to deal with `fields` related to that. And definitely not trying to Kotlinize any of those classes, as this will then force me to enter this rabbit hole anyway. Otherwise, as you can understand, this will get quickly out of hand, definitely out of scope. 🤷

ParaskP7 replied:
FYI: This is now done: 69601155de97c4bcce6fbe4dbba18cb3beea73d3

mkevins replied:
> I would still avoid adding the extra Objects.requireNonNull(mNoteList.get(position), "Note cannot be null") check there, because that would mean that this is now becoming the rule here, and we should strive doing that on every such instance of Fragment getItem(int position) { ... } in our project.

I don't agree that adding this assertion means that it is "becoming a rule", and that doing so requires adding a similar check everywhere that something _appears_ similar throughout the codebase. This kind of rationale would prevent us from making small changes toward a larger goal. I think we should focus on this instance, and avoid prematurely generalizing.

>  But, I am not convinced yet that it is actually needed, I need to investigate if there are cases where getItem(...) will be called on the adapter but the list to get that item from might be empty, or else not having that item on that exact position, or else if that item can be null (so far my assumption is no to all of those questions). Does that make sense?

I don't think any deep investigation is needed in order to make an assertion. In fact, without making an _explicit_ assertion, we are making an _implicit_ one. But, the trace will not be friendly in case of a crash. The point here is that `mNoteList` being an `ArrayList` can have any element as `null`, even when it is not empty. In Java, when considering an `ArrayList<T>`, we not only must consider the arrayList itself as `@Nullable`, but the collection items as well (i.e. `T` is `@Nullable`). Rather than chasing down every possible insertion / mutation of this list so prove to ourselves it might be `@NonNull` (which could be invalidated by later code changes anyway), the suggestion I'm making is that we do _not_ try to go that deep.

Given that we _know_ that `mNoteList.get(position)` might be `null`, and that `getDetailFragmentForNote` accepts only `@NonNull` arguments, it is proper to do this assertion _before_ invoking the annotated method. Otherwise, we get this confusing situation where the trace points us to [this line](https://github.com/wordpress-mobile/WordPress-Android/commit/69601155de97c4bcce6fbe4dbba18cb3beea73d3#diff-e48edd5a3728c682d33322513d5ae0d968c321d8cda405d1dd45e9f4ac8b4152R404) for a `NullPointerException`, the `@NonNull` parameter annotation is clearly violated, and whoever is analyzing this crash has to work backwards and re-establish the actual source of the unexpected nullness, though we have already worked that out. I'd rather _fail-fast_ here, conveying our knowledge to the would-be crash analyst.

ParaskP7 replied:
👋 @mkevins and let me preface this response of mine with the fact that I don't feel strongly about adding `Objects.requireNonNull(mNoteList.get(position)`, nor it is hard for me to go and add this commit, it will take me just a minute, I just want to understand your points to the fullest so I can be more efficient going forward. 🙏 

-----

> I don't agree that adding this assertion means that it is "becoming a rule", and that doing so requires adding a similar check everywhere that something appears similar throughout the codebase.

First of all thanks for disagreeing with me Matt , this means you care and that you want the best possible outcome from this PR, and in general that you care for the overall success of the project, much appreciated! 🥇 

> This kind of rationale would prevent us from making small changes toward a larger goal.

I understand that and can agree with you on it. 💯

>  I think we should focus on this instance, and avoid prematurely generalizing.

Having said the about, I want to share with you my rational and how I ended-up with this generalizing, not being sure if that is premature or not, let me know if you still believe so, I would really appreciate your feedback here:

1. I first searched for `override fun getItem\(position\: Int\)\: Fragment \{` and this got me the below list of usages:
    1. [CommentDetailFragmentAdapter.java](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragmentAdapter.java#L36-L40) -> `getComment(position)` -> if check on `position` -> `mCommentList.get(position)` 🤔
    2. [HistoryDetailContainerFragment.java](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/history/HistoryDetailContainerFragment.java#L356-L359) -> `mRevisions.get(position)` 🎯
    3. [MediaPreviewActivity.java](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/media/MediaPreviewActivity.java#L395-L423) -> `mMediaIdOrUrlList.get(position)` 🎯
    4. [EditPostActivity.java](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L2361-L2412) -> switch case on `position` 🤔
    5. [ReaderPhotoViewerActivity.java](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPhotoViewerActivity.java#L283-L288) -> `mImageList.get(position)` 🎯 
    6. [ReaderPostPagerActivity.java](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostPagerActivity.java#L1012-L1027) -> if check on `position` -> `mIdList.get(position)` 🤔
    7. [ReaderSubsActivity.java](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderSubsActivity.java#L548-L551) -> `mFragments.get(position)` 🎯 
2. I then searched for: `override fun getItem\(position\: Int\)\: Fragment \{`
    1. [PagesFragment.kt](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/pages/PagesFragment.kt#L663-L667) -> `pageTypes[position]` 🎯
    2. [SubfilterPageFragment.kt](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/reader/subfilter/SubfilterPageFragment.kt#L158-L162) -> `filterCategory[position]` 🎯

🎯 -> I think it is safe to assume that this fits our case, thus can be consider generalizing.
🤔 -> Maybe, doesn't fit to our case, thus can be considered an exception to the rule, but I am not 100% sure without making a deeper dive.

> I don't think any deep investigation is needed in order to make an assertion. In fact, without making an explicit assertion, we are making an implicit one. But, the trace will not be friendly in case of a crash. 

I agree with that and the trace being not friendly. Yet, do you think we should strive to get a friendly NPE trace in every such case? I am just trying to find the pattern here so I can apply it to my future PRs.

> The point here is that mNoteList being an ArrayList can have any element as null, even when it is not empty. In Java, when considering an ArrayList<T>, we not only must consider the arrayList itself as @Nullable, but the collection items as well (i.e. T is @Nullable).

I understand the technicalities here for a general Java based `ArrayList<T>`, but do you think this applies to `FragmentStatePagerAdapter`, `FragmentPagerAdapter`, `PagerAdapter` and its API contract? 🤔

> Rather than chasing down every possible insertion / mutation of this list so prove to ourselves it might be @NonNull (which could be invalidated by later code changes anyway), the suggestion I'm making is that we do not try to go that deep.

I agree here and I would suggest the same.

> Given that we know that mNoteList.get(position) might be null, and that getDetailFragmentForNote accepts only @NonNull arguments, it is proper to do this assertion before invoking the annotated method.

Ah, please help me understand this bit, how do we know that `mNoteList.get(position)` might be null? Do you mean we know that because of Java and the fact that `ArrayList<T>` can accept nullable items, or because you indeed have identified cases when the `FragmentStatePagerAdapter`, `FragmentPagerAdapter`, `PagerAdapter` and its API can produce such case, possibly valid cases? 🤔

> Otherwise, we get this confusing situation where the trace points us to [this line](https://github.com/wordpress-mobile/WordPress-Android/commit/69601155de97c4bcce6fbe4dbba18cb3beea73d3#diff-e48edd5a3728c682d33322513d5ae0d968c321d8cda405d1dd45e9f4ac8b4152R404) for a NullPointerException, the @NonNull parameter annotation is clearly violated, and whoever is analyzing this crash has to work backwards and re-establish the actual source of the unexpected nullness, though we have already worked that out.

I agree this is confusing.

> I'd rather fail-fast here, conveying our knowledge to the would-be crash analyst.

👍

mkevins replied:
> First of all thanks for disagreeing with me Matt , this means you care and that you want the best possible outcome from this PR, and in general that you care for the overall success of the project, much appreciated! 🥇

Of course, and I feel the same way toward your position on these discussions. 👍 

> I agree with that and the trace being not friendly. Yet, do you think we should strive to get a friendly NPE trace in every such case? I am just trying to find the pattern here so I can apply it to my future PRs.

> I understand the technicalities here for a general Java based ArrayList<T>, but do you think this applies to FragmentStatePagerAdapter, FragmentPagerAdapter, PagerAdapter and its API contract? 🤔

No, I think it is quite possible that there exists a water-tight argument that the item in these other cases (and even possibly in _this_ case) is guaranteed to be `@NonNull`. But, we have not annotated it as such, so the suggestion I'm making is purely a shallow suggestion to improve a trace (albeit one that may never occur).

> Ah, please help me understand this bit, how do we know that mNoteList.get(position) might be null? Do you mean we know that because of Java and the fact that ArrayList<T> can accept nullable items, or because you indeed have identified cases when the FragmentStatePagerAdapter, FragmentPagerAdapter, PagerAdapter and its API can produce such case, possibly valid cases? 🤔

The first one. Since we aren't strictly taking about framework APIs (i.e. this nullity is governed by our own code, it is _possible_ we haven't considered every case).

From the discussion, I wonder if you consider adding `Objects.requireNotNull` as some equivalent to annotating as `@Nullable`? 🤔  I don't feel strongly enough about adding it to block this PR (hence why already approved 😄 ), but this discussion is useful anyway, imo, so I'll explain how I'm thinking about it:

I see `@Nullable` and `@NonNull` annotations as trail markers, delineating the boundaries between the world of `null` and the world of no more NPEs. In the world of `null`, it's quite possible to have no NPEs, but requires constant diligence on behalf of developers, and often a defensive programming mindset that can lead to redundancy in guard code, potential dead code, and more complexity due to branching. In the world of `@NonNull`, NPEs are not even possible, and our code can be elegant, expressive, and concise.

Imo, it would be great to shrink the `@Nullable` world down to as small a surface area as possible, and expand the `@NonNull` area. We are like park rangers, maintaining the trail markers, and when we want to convert an area to `@NonNull`, we typically have to choose between gracefully guarding, or checking and crashing. To me, `requireNotNull` acts as a gateway to allow us to connect the unmarked trails that we encounter with our growing network of `@NonNull` trails. If we are going to crash, we know exactly where.

All that said, I think I understand where you are coming from in that this is like a "model" PR, and I understand your hesitation to include an "ad hoc" check.

ParaskP7 replied:
> Of course, and I feel the same way toward your position on these discussions. 👍

❤️ 🥊

> No, I think it is quite possible that there exists a water-tight argument that the item in these other cases (and even possibly in this case) is guaranteed to be @NonNull. But, we have not annotated it as such, so the suggestion I'm making is purely a shallow suggestion to improve a trace (albeit one that may never occur).

👍

> The first one. Since we aren't strictly taking about framework APIs (i.e. this nullity is governed by our own code, it is possible we haven't considered every case).

👍

> From the discussion, I wonder if you consider adding Objects.requireNotNull as some equivalent to annotating as @Nullable? 🤔 I don't feel strongly enough about adding it to block this PR (hence why already approved 😄 ), but this discussion is useful anyway, imo, so I'll explain how I'm thinking about it:

👍

> I see @Nullable and @NonNull annotations as trail markers, delineating the boundaries between the world of null and the world of no more NPEs. In the world of null, it's quite possible to have no NPEs, but requires constant diligence on behalf of developers, and often a defensive programming mindset that can lead to redundancy in guard code, potential dead code, and more complexity due to branching. In the world of @NonNull, NPEs are not even possible, and our code can be elegant, expressive, and concise.

I agree! 💯

> Imo, it would be great to shrink the @Nullable world down to as small a surface area as possible, and expand the @NonNull area. We are like park rangers, maintaining the trail markers, and when we want to convert an area to @NonNull, we typically have to choose between gracefully guarding, or checking and crashing. To me, requireNotNull acts as a gateway to allow us to connect the unmarked trails that we encounter with our growing network of @NonNull trails. If we are going to crash, we know exactly where.

This is a nice way of thinking about it, like! 🌟

> All that said, I think I understand where you are coming from in that this is like a "model" PR, and I understand your hesitation to include an "ad hoc" check.

This, plus the fact that I still want to do a deeper dive in the world of `List` and how we should best deal with items in a list, always in terms of their nullability.

Thanks for your understanding Matt, I'll now proceed with merging this PR as is, but I feel all this discussion on this matter was very worthy. From now on I'll be much more sensitive to using this mechanism and this way switch from a nullable to a non-null world, it's a beautiful world after all! 🚀

mkevins reviewed (approved):
Nice work Petros! I tested this (on a Pixel 3a physical device), and everything is working as expected. I also reviewed the code, and everything looks good there as well. I left a few comments, but none of them are blocking, imo.

If possible, in future PRs of this kind, if it is simple enough, it may be nice to convert to Kotlin along with the change to `viewBinding`, since we can utilize `lateinit` allowing the binding to be a non-null type.

ParaskP7 commented:
> Nice work Petros! I tested this (on a Pixel 3a physical device), and everything is working as expected. I also reviewed the code, and everything looks good there as well. I left a few comments, but none of them are blocking, imo.

Awesome, thank you so much for reviewing and testing this PR @mkevins , you rock! 🙇 ❤️ 🚀 

> If possible, in future PRs of this kind, if it is simple enough, it may be nice to convert to Kotlin along with the change to viewBinding, since we can utilize lateinit allowing the binding to be a non-null type.

I am not sure about mixing Kotlinization with `ViewBinding` work. I think it is better to keep the PR simple and not mix the too of them, as they both serve a different purpose. Also, although Kotlinizing a small activity/fragment class might seems easy enough, doing the same for a class of more than 100 lines of code might not be that trivial.

PS: When I mean Kotlinization, the way at least I do it, is to do it the right way and till the end, not just depend on the automated IDE conversion, but also, taking it further with removing any `!!` leftovers, fixing any warnings (compile and Detekt ones) and making the code as idiomatic as possible (applying some common sense of course, not trying to get to the other side of having an idiomatic code).

ParaskP7 committed:
Analysis: Add missing nn-a to get detail fragment for note method

This is done in order for the 'getDetailFragmentForNote(...)' method and
its return type to adhere to the 'getItem(...)' method and its
'@NonNull' return type.

mkevins commented:
> I am not sure about mixing Kotlinization with `ViewBinding` work. I think it is better to keep the PR simple and not mix the too of them, as they both serve a different purpose. Also, although Kotlinizing a small activity/fragment class might seems easy enough, doing the same for a class of more than 100 lines of code might not be that trivial.

I agree (e.g. my meaning when I mentioned "_if it is simple enough_").

> PS: When I mean Kotlinization, the way at least I do it, is to do it the right way and till the end, not just depend on the automated IDE conversion, but also, taking it further with removing any `!!` leftovers, fixing any warnings (compile and Detekt ones) and making the code as idiomatic as possible (applying some common sense of course, not trying to get to the other side of having an idiomatic code).

Of course, and I don't think just mindlessly using the automated IDE conversion should ever be an acceptable practice. 😄 My main point, here, is that much of the work is overlapping and has synergies. In particular, the "removing and `!!` leftovers" relates directly with what we are doing here with null analysis.

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-Android into refactor/replace-findviewbyid-with-viewbinding-for-notifications
